src/cargo/core/resolver/mod.rs

```diff
@@ -247,158 +193,245 @@ pub fn resolve(summary: &Summary, method: &Method,
         activations: HashMap::new(),
         visited: HashSet::new(),
-    });
+    };
     let _p = profile::start(format!("resolving: {}", summary.package_id()));
-    match try!(activate(cx, registry, &summary, method, &mut |cx, _| Ok(Ok(cx)))) {
-        Ok(cx) => {
-            debug!("resolved: {:?}", cx.resolve);
-            Ok(cx.resolve)
-        }
-        Err(e) => Err(e),
-    }
+    activate_deps_loop(cx, registry, summary, method)
 }
 
 /// Attempts to activate the summary `parent` in the context `cx`.
 ///
 /// This function will pull dependency summaries from the registry provided, and
 /// the dependencies of the package will be determined by the `method` provided.
-/// Once the resolution of this package has finished **entirely**, the current
-/// context will be passed to the `finished` callback provided.
-fn activate(mut cx: Box<Context>,
+/// If `parent` was activated, this function returns the dependency frame to
+/// iterate through next.
+fn activate(cx: &mut Context,
             registry: &mut Registry,
-            parent: &Rc<Summary>,
-            method: &Method,
-            finished: &mut FnMut(Box<Context>, &mut Registry) -> ResolveResult)
-            -> ResolveResult {
+            parent: Rc<Summary>,
+            method: &Method)
+            -> CargoResult<Option<DepsFrame>> {
     // Dependency graphs are required to be a DAG, so we keep a set of
     // packages we're visiting and bail if we hit a dupe.
-    let id = parent.package_id();
+    let id = parent.package_id().clone();
     if !cx.visited.insert(id.clone()) {
         return Err(human(format!("cyclic package dependency: package `{}` \
                                   depends on itself", id)))
     }
 
     // If we're already activated, then that was easy!
-    if cx.flag_activated(parent, method) {
-        cx.visited.remove(id);
-        return finished(cx, registry)
+    if cx.flag_activated(&parent, method) {
+        cx.visited.remove(&id);
+        return Ok(None);
     }
-    trace!("activating {}", parent.package_id());
+    debug!("activating {}", parent.package_id());
 
-    let deps = try!(cx.build_deps(registry, parent, method));
+    let deps = try!(cx.build_deps(registry, &parent, method));
 
-    activate_deps(cx, registry, parent, deps.iter(), 0, &mut |mut cx, registry| {
-        cx.visited.remove(id);
-        finished(cx, registry)
-    })
+    Ok(Some(DepsFrame{
+        parent: parent,
+        remaining_siblings: RcVecIter::new(deps),
+        id: id,
+    }))
 }
 
-/// Activates the dependencies for a package, one by one in turn.
-///
-/// This function will attempt to activate all possible candidates for each
-/// dependency of the package specified by `parent`. The `deps` iterator
-/// provided is an iterator over all dependencies where each element yielded
-/// informs us what the candidates are for the dependency in question.
+#[derive(Clone)]
+struct RcVecIter<Elem> {
```

Comment:
Conventionally type parameters like this are typically called something like `T` instead of `Elem` (same below as well)